### PR TITLE
Add setter functions for email message

### PR DESCRIPTION
**Pull request by michvllni**
per #16149

Adding the following setter functions to the email message codeunit.
- SetBody
- AppendToBody
- SetSubject
- SetBodyHTMLFormatted
- SetRecipients (Both List of Text and Text)

Also renaming the previous setter functions in the codeunit "email message impl" to for example setBodyValue to not have conflicts with the naming in the "email message" codeunit

### CODE REVIEW

File: Modules/System/Email/src/Message/EmailMessage.Codeunit.al [hunk 1]
```diff
@@ -78,6 +78,24 @@ codeunit 8904 "Email Message"
         exit(EmailMessageImpl.GetBody());
     end;
 
+    /// <summary>
+    /// Sets the body of the email message.
+    /// /// <param name="Body">The body to set to the email message.</param>
```
Comment (mazhelez) [inline Modules/System/Email/src/Message/EmailMessage.Codeunit.al:83]:
```suggestion
    /// <param name="Body">The body to set to the email message.</param>
```

File: Modules/System/Email/src/Message/EmailMessage.Codeunit.al [hunk 2]
```diff
@@ -78,6 +78,24 @@ codeunit 8904 "Email Message"
...
+    /// </summary>
+    procedure SetBody(Body: Text)
+    begin
+        EmailMessageImpl.SetBody(Body);
+    end;
+
+    /// <summary>
+    /// Appends to the body of the email message.
+    /// /// <param name="Value">The value to append to the body of the email message.</param>
```
Comment (mazhelez) [inline Modules/System/Email/src/Message/EmailMessage.Codeunit.al:92]:
```suggestion
    /// <param name="Value">The value to append to the body of the email message.</param>
```

File: Modules/System/Email/src/Message/EmailMessage.Codeunit.al [hunk 3]
```diff
@@ -87,6 +105,15 @@ codeunit 8904 "Email Message"
         exit(EmailMessageImpl.GetSubject());
     end;
 
+    /// <summary>
+    /// Sets the subject of the email message.
+    /// /// <param name="Subject">The subject to set to the email message.</param>
```
Comment (mazhelez) [inline Modules/System/Email/src/Message/EmailMessage.Codeunit.al:110]:
```suggestion
    /// <param name="Subject">The subject to set to the email message.</param>
```

File: Modules/System/Email/src/Message/EmailMessage.Codeunit.al [hunk 4]
```diff
@@ -96,6 +123,15 @@ codeunit 8904 "Email Message"
         exit(EmailMessageImpl.IsBodyHTMLFormatted());
     end;
 
+    /// <summary>
+    /// Sets whether the email body is formatted in HTML.
+    /// /// <param name="Value">True if the email body is formatted in HTML; otherwise - false.</param>
```
Comment (mazhelez) [inline Modules/System/Email/src/Message/EmailMessage.Codeunit.al:128]:
```suggestion
    /// <param name="Value">True if the email body is formatted in HTML; otherwise - false.</param>
```

File: Modules/System/Email/src/Message/EmailMessage.Codeunit.al [hunk 5]
```diff
@@ -115,6 +151,26 @@ codeunit 8904 "Email Message"
         Recipients := EmailMessageImpl.GetRecipients(RecipientType);
     end;
 
+    /// <summary>
+    /// Sets the recipents of a certain type of the email message.
+    /// </summary>
+    /// <param name="RecipientType">Specifies the type of the recipients.</param>
+    /// <param name="Recipients">Specifies the list of the recipients' email addresses as a semicolon (;) separated list.</param>
+    procedure SetRecipients(RecipientType: Enum "Email Recipient Type"; var Recipients: Text)
```
Comment (mazhelez) [inline Modules/System/Email/src/Message/EmailMessage.Codeunit.al:159]: There's no need for the `var` here.

File: Modules/System/Email/src/Message/EmailMessage.Codeunit.al [hunk 6]
```diff
@@ -115,6 +151,26 @@ codeunit 8904 "Email Message"
...
+    begin
+        EmailMessageImpl.SetRecipients(RecipientType, Recipients);
+    end;
+
+    /// <summary>
+    /// Sets the recipents of a certain type of the email message.
+    /// </summary>
+    /// <param name="RecipientType">Specifies the type of the recipients.</param>
+    /// <param name="Recipients">Specifies the list of the recipients' email addresses.</param>
+    procedure SetRecipients(RecipientType: Enum "Email Recipient Type"; var Recipients: list of [Text])
```
Comment (mazhelez) [inline Modules/System/Email/src/Message/EmailMessage.Codeunit.al:169]: Also here... no need for the `var`

**Review by mazhelez — COMMENTED** [review]
Super! Great to see you took the matter in your hands.

Just to be thorough, you can add some test. Especially for `AppendToBody`.

File: Modules/System/Email/src/Message/EmailMessage.Codeunit.al (hunk 5 above)
Comment (michvllni) [reply Modules/System/Email/src/Message/EmailMessage.Codeunit.al]: You are correct, it was late when I copied those :-) I resolved them

**Comment by michvllni** [comment]
> Super! Great to see you took the matter in your hands.
> 
> Just to be thorough, you can add some test. Especially for `AppendToBody`.

You're completely right, I added some tests. I hope they are thorough enough

File: Modules/System Tests/Email/src/EmailMessageUnitTest.Codeunit.al [hunk 7]
```diff
@@ -522,6 +522,157 @@ codeunit 134689 "Email Message Unit Test"
         Assert.ExpectedError(EmailMessageSentCannotDeleteRecipientErr);
     end;
 
+    [Test]
+    [TransactionModel(TransactionModel::AutoRollback)]
+    procedure SetBodyTest()
+    var
+        Message: Codeunit "Email Message";
+        Recipients: List of [Text];
+        Result: List of [Text];
```
Comment (mazhelez) [inline Modules/System Tests/Email/src/EmailMessageUnitTest.Codeunit.al:531]: There are some unused variables here ;)

File: Modules/System Tests/Email/src/EmailMessageUnitTest.Codeunit.al [hunk 8]
```diff
@@ -522,6 +522,157 @@ codeunit 134689 "Email Message Unit Test"
...
+        Index: Integer;
+    begin
+        // Initialize
+        PermissionsMock.Set('Email Edit');
+
+        // Exercise
+        Message.Create(Recipients, 'Test subject', 'Test body', true);
```
Comment (mazhelez) [inline Modules/System Tests/Email/src/EmailMessageUnitTest.Codeunit.al:538]: The creation part belongs to the "initialize" section.

File: Modules/System Tests/Email/src/EmailMessageUnitTest.Codeunit.al [hunk 9]
```diff
@@ -522,6 +522,157 @@ codeunit 134689 "Email Message Unit Test"
...
+        Message.AppendToBody(' extended');
```
Comment (mazhelez) [inline Modules/System Tests/Email/src/EmailMessageUnitTest.Codeunit.al:558]: At least one more test with apeending an empty string, please

File: Modules/System Tests/Email/src/EmailMessageUnitTest.Codeunit.al [hunk 10]
```diff
@@ -522,6 +522,157 @@ codeunit 134689 "Email Message Unit Test"
...
+        Message.Create(Recipients, 'Test subject', 'Test body', true);
+        Recipients.Add('recipient1@test.com');
+        Recipients.Add('recipient2@test.com');
+        Recipients.Add('recipient3@test.com');
+        Message.SetRecipients(Enum::"Email Recipient Type"::"To", Recipients);
+        Message.GetRecipients(Enum::"Email Recipient Type"::"To", Result);
+
+        // Verify
+        for Index := 1 to Result.Count() do
+            Assert.AreEqual(StrSubstNo(RecipientLbl, Index), Result.Get(Index), 'A different recipient was expected');
```
Comment (mazhelez) [inline Modules/System Tests/Email/src/EmailMessageUnitTest.Codeunit.al:623]: I'd also like to see a test that sets the CC recipients (for instance) and checks that To and BCC are not changed.

File: Modules/System/Email/src/Message/EmailMessage.Codeunit.al [hunk 11]
```diff
@@ -115,6 +151,26 @@ codeunit 8904 "Email Message"
...
+    begin
+        EmailMessageImpl.SetRecipients(RecipientType, Recipients);
+    end;
+
+    /// <summary>
+    /// Sets the recipents of a certain type of the email message.
+    /// </summary>
+    /// <param name="RecipientType">Specifies the type of the recipients.</param>
+    /// <param name="Recipients">Specifies the list of the recipients' email addresses.</param>
+    procedure SetRecipients(RecipientType: Enum "Email Recipient Type"; Recipients: list of [Text])
```
Comment (mazhelez) [inline Modules/System/Email/src/Message/EmailMessage.Codeunit.al:169]: Just food for thought: isn't `AddRecipient(RecipientType: Enum "Email Recipient Type"; Recipient: Text)` a bit more useful? Or at least useful as much?

Is there a scenario where the whole list of recipients will be changed?

**Review by mazhelez — CHANGES_REQUESTED** [review]
A few more comments to address and we'll be ready to uptake the PR!

**Comment by mazhelez** [comment]
A test is failing:

```
[2022-02-17 22:36:26] Name: Email Message Unit Test:AppendToBodyTest
Method: AppendToBodyTest
Time: 0.024
Message: 
Assert.AreEqual failed. Expected:<Test body extended> (Text). Actual:< extended> (Text). A different body was expected.
StackTrace: 
Library Assert(CodeUnit 130002).AreEqual line 3 - Library Assert by Microsoft
Email Message Unit Test(CodeUnit 134689).AppendToBodyTest line 12 - Email Test by Microsoft
Test Runner - Mgt(CodeUnit 130454).RunTests line 18 - Test Runner by Microsoft
Test Runner - Isol. Codeunit(CodeUnit 130450).OnRun(Trigger) line 4 - Test Runner by Microsoft
Test Suite Mgt.(CodeUnit 130456).RunTests line 2 - Test Runner by Microsoft
Test Suite Mgt.(CodeUnit 130456).RunSelectedTests line 35 - Test Runner by Microsoft
Test Suite Mgt.(CodeUnit 130456).RunNextTest line 16 - Test Runner by Microsoft
Command Line Test Tool(Page 130455).RunNextTest - OnAction(Trigger) line 7 - Test Runner by Microsoft
```

File: Modules/System/Email/src/Message/EmailMessage.Codeunit.al (hunk 11 above)
Comment (michvllni) [reply Modules/System/Email/src/Message/EmailMessage.Codeunit.al:179]: Good idea, I've added it

**Comment by michvllni** [comment]
I fixed your comments. I hope everything works as expected now

File: Modules/System/Email/src/Message/EmailMessageImpl.Codeunit.al [hunk 12]
```diff
@@ -355,6 +355,36 @@ codeunit 8905 "Email Message Impl."
...
     end;
 
+    procedure AddRecipient(RecipientType: Enum "Email Recipient Type"; Recipient: Text)
+    var
+        EmailRecipientRecord: Record "Email Recipient";
+        UniqueRecipients: Dictionary of [Text, Text];
+        Recipients: List of [Text];
+        tmpRecipient: Text;
+    begin
+        if Recipient <> '' then
```
Comment (Groenbech96) [inline Modules/System/Email/src/Message/EmailMessageImpl.Codeunit.al:368]: This means that you do not add anything if the recipient is not empty.
Is that on purpose?

File: Modules/System Tests/Email/src/EmailMessageUnitTest.Codeunit.al [hunk 13]
```diff
@@ -631,61 +633,157 @@ codeunit 134689 "Email Message Unit Test"
...
+        // Exercise
+        Message.Create(ResultTo, 'Test subject', 'Test body', true, ResultCc, ResultBCc);
+        Recipient := 'recipient1@test.com';
+        Message.AddRecipient(Enum::"Email Recipient Type"::"To", Recipient);
+        Message.GetRecipients(Enum::"Email Recipient Type"::"To", ResultTo);
+        Message.GetRecipients(Enum::"Email Recipient Type"::"Cc", ResultCc);
+        Message.GetRecipients(Enum::"Email Recipient Type"::"BCc", ResultBCc);
+
+        // Verify
+        Assert.AreEqual(Recipient, ResultTo.Get(0), 'A different recipient was expected');
```
Comment (Groenbech96) [inline Modules/System Tests/Email/src/EmailMessageUnitTest.Codeunit.al:730]: Al is 1 indexed.
This will fail

File: Modules/System Tests/Email/src/EmailMessageUnitTest.Codeunit.al [hunk 14]
```diff
@@ -631,61 +633,157 @@ codeunit 134689 "Email Message Unit Test"
...
+        Message.Create(ResultTo, 'Test subject', 'Test body', true, ResultCc, ResultBCc);
+        Recipient := 'recipient1@test.com';
+        Message.AddRecipient(Enum::"Email Recipient Type"::"Cc", Recipient);
+        Message.GetRecipients(Enum::"Email Recipient Type"::"To", ResultTo);
+        Message.GetRecipients(Enum::"Email Recipient Type"::"Cc", ResultCc);
+        Message.GetRecipients(Enum::"Email Recipient Type"::"BCc", ResultBCc);
+
+        // Verify
+        Assert.AreEqual(ResultTo.Count, 0, 'Zero To recipients were expected');
+        Assert.AreEqual(Recipient, ResultCc.Get(0), 'A different recipient was expected');
```
Comment (Groenbech96) [inline Modules/System Tests/Email/src/EmailMessageUnitTest.Codeunit.al:758]: Same here :)

File: Modules/System Tests/Email/src/EmailMessageUnitTest.Codeunit.al [hunk 15]
```diff
@@ -631,61 +633,157 @@ codeunit 134689 "Email Message Unit Test"
...
+        Recipient := 'recipient1@test.com';
+        Message.AddRecipient(Enum::"Email Recipient Type"::"BCc", Recipient);
+        Message.GetRecipients(Enum::"Email Recipient Type"::"To", ResultTo);
+        Message.GetRecipients(Enum::"Email Recipient Type"::"Cc", ResultCc);
+        Message.GetRecipients(Enum::"Email Recipient Type"::"BCc", ResultBCc);
+
+        // Verify
+        Assert.AreEqual(ResultTo.Count, 0, 'Zero To recipients were expected');
+        Assert.AreEqual(ResultCc.Count, 0, 'Zero Cc recipients were expected');
+        Assert.AreEqual(Recipient, ResultBCc.Get(0), 'A different recipient was expected');
```
Comment (Groenbech96) [inline Modules/System Tests/Email/src/EmailMessageUnitTest.Codeunit.al:786]: Same :)

**Review by Groenbech96 — CHANGES_REQUESTED** [review]
Minor things need to be fixed for the PR. :)

File: Modules/System/Email/src/Message/EmailMessageImpl.Codeunit.al [hunk 16]
```diff
@@ -324,6 +358,36 @@ codeunit 8905 "Email Message Impl."
...
+            exit;
+        EmailRecipientRecord.SetRange("Email Message Id", Message.Id);
+        EmailRecipientRecord.SetRange("Email Recipient Type", RecipientType);
+
+        if not EmailRecipientRecord.IsEmpty() then
+            EmailRecipientRecord.DeleteAll();
```
Comment (mazhelez) [inline Modules/System/Email/src/Message/EmailMessageImpl.Codeunit.al:374]: Why deleting the existing recipients? That shouldn't be okay.

File: Modules/System/Email/src/Message/EmailMessageImpl.Codeunit.al (hunk 12 above)
Comment (michvllni) [reply Modules/System/Email/src/Message/EmailMessageImpl.Codeunit.al:368]: Oops, I meant of course to exit if the recipient is empty, not if it is not. Good Catch!

File: Modules/System Tests/Email/src/EmailMessageUnitTest.Codeunit.al (hunk 13 above)
Comment (michvllni) [reply Modules/System Tests/Email/src/EmailMessageUnitTest.Codeunit.al:730]: Fixed it

File: Modules/System Tests/Email/src/EmailMessageUnitTest.Codeunit.al (hunk 14 above)
Comment (michvllni) [reply Modules/System Tests/Email/src/EmailMessageUnitTest.Codeunit.al:758]: Fixed it

File: Modules/System Tests/Email/src/EmailMessageUnitTest.Codeunit.al (hunk 15 above)
Comment (michvllni) [reply Modules/System Tests/Email/src/EmailMessageUnitTest.Codeunit.al:786]: Fixed it

File: Modules/System/Email/src/Message/EmailMessageImpl.Codeunit.al (hunk 16 above)
Comment (michvllni) [reply Modules/System/Email/src/Message/EmailMessageImpl.Codeunit.al:374]: You're absolutely right. I made a mistake when copying from the SetRecipients procedure, I've resolved it

File: Modules/System/Email/src/Message/EmailMessage.Codeunit.al [hunk 17]
```diff
@@ -78,6 +78,24 @@ codeunit 8904 "Email Message"
         exit(EmailMessageImpl.GetBody());
     end;
 
+    /// <summary>
+    /// Sets the body of the email message.
+    /// <param name="Body">The body to set to the email message.</param>
```
Comment (mazhelez) [inline Modules/System/Email/src/Message/EmailMessage.Codeunit.al:83]: _Param_ tag should be outside of _summary_.

File: Modules/System/Email/src/Message/EmailMessage.Codeunit.al [hunk 18]
```diff
@@ -78,6 +78,24 @@ codeunit 8904 "Email Message"
...
+    /// </summary>
+    procedure SetBody(Body: Text)
+    begin
+        EmailMessageImpl.SetBody(Body);
+    end;
+
+    /// <summary>
+    /// Appends to the body of the email message.
+    /// <param name="Value">The value to append to the body of the email message.</param>
```
Comment (mazhelez) [inline Modules/System/Email/src/Message/EmailMessage.Codeunit.al:92]: _Param_ tag should be outside of _summary_.

File: Modules/System/Email/src/Message/EmailMessage.Codeunit.al [hunk 19]
```diff
@@ -87,6 +105,15 @@ codeunit 8904 "Email Message"
         exit(EmailMessageImpl.GetSubject());
     end;
 
+    /// <summary>
+    /// Sets the subject of the email message.
+    /// <param name="Subject">The subject to set to the email message.</param>
```
Comment (mazhelez) [inline Modules/System/Email/src/Message/EmailMessage.Codeunit.al:110]: _Param_ tag should be outside of _summary_.

File: Modules/System/Email/src/Message/EmailMessage.Codeunit.al [hunk 20]
```diff
@@ -96,6 +123,15 @@ codeunit 8904 "Email Message"
         exit(EmailMessageImpl.IsBodyHTMLFormatted());
     end;
 
+    /// <summary>
+    /// Sets whether the email body is formatted in HTML.
+    /// <param name="Value">True if the email body is formatted in HTML; otherwise - false.</param>
```
Comment (mazhelez) [inline Modules/System/Email/src/Message/EmailMessage.Codeunit.al:128]: _Param_ tag should be outside of _summary_.

File: Modules/System/Email/src/Message/EmailMessage.Codeunit.al [hunk 21]
```diff
@@ -115,6 +151,36 @@ codeunit 8904 "Email Message"
         Recipients := EmailMessageImpl.GetRecipients(RecipientType);
     end;
 
+    /// <summary>
+    /// Sets the recipents of a certain type of the email message.
```
Comment (mazhelez) [inline Modules/System/Email/src/Message/EmailMessage.Codeunit.al:155]:
```suggestion
    /// Sets the recipient of a certain type of the email message.
```

File: Modules/System/Email/src/Message/EmailMessage.Codeunit.al [hunk 22]
```diff
@@ -115,6 +151,36 @@ codeunit 8904 "Email Message"
...
+    /// </summary>
+    /// <param name="RecipientType">Specifies the type of the recipients.</param>
+    /// <param name="Recipients">Specifies the list of the recipients' email addresses as a semicolon (;) separated list.</param>
+    procedure SetRecipients(RecipientType: Enum "Email Recipient Type"; Recipients: Text)
+    begin
+        EmailMessageImpl.SetRecipients(RecipientType, Recipients);
+    end;
+
+    /// <summary>
+    /// Adds a recipent of a certain type to the email message.
```
Comment (mazhelez) [inline Modules/System/Email/src/Message/EmailMessage.Codeunit.al:165]:
```suggestion
    /// Adds a recipient of a certain type to the email message.
```

File: Modules/System/Email/src/Message/EmailMessage.Codeunit.al [hunk 23]
```diff
@@ -115,6 +151,36 @@ codeunit 8904 "Email Message"
...
+    /// </summary>
+    /// <param name="RecipientType">Specifies the type of the recipient.</param>
+    /// <param name="Recipient">Specifies the recipient's email address.</param>
+    procedure AddRecipient(RecipientType: Enum "Email Recipient Type"; Recipient: Text)
+    begin
+        EmailMessageImpl.AddRecipient(RecipientType, Recipient);
+    end;
+
+    /// <summary>
+    /// Sets the recipents of a certain type of the email message.
```
Comment (mazhelez) [inline Modules/System/Email/src/Message/EmailMessage.Codeunit.al:175]:
```suggestion
    /// Sets the recipient of a certain type of the email message.
```

File: Modules/System/Email/src/Message/EmailMessageImpl.Codeunit.al [hunk 24]
```diff
@@ -111,26 +111,60 @@ codeunit 8905 "Email Message Impl."
...
+        SetBodyValue(BodyText);
+        Modify();
+    end;
+
+    procedure AppendToBody(BodyText: Text)
+    var
+        BodyOutStream: OutStream;
+        Body: Text;
+    begin
+        if BodyText = '' then
```
Comment (mazhelez) [inline Modules/System/Email/src/Message/EmailMessageImpl.Codeunit.al:125]: What about something even shorter:

```
        if BodyText = '' then
            exit;

        ExistingBodyText := GetBody();
        SetBody(ExistingBodyText + BodyText);
```

File: Modules/System/Email/src/Message/EmailMessageImpl.Codeunit.al [hunk 25]
```diff
@@ -324,6 +358,31 @@ codeunit 8905 "Email Message Impl."
...
     end;
 
+    procedure AddRecipient(RecipientType: Enum "Email Recipient Type"; Recipient: Text)
+    var
+        EmailRecipientRecord: Record "Email Recipient";
+        UniqueRecipients: Dictionary of [Text, Text];
+        Recipients: List of [Text];
+        tmpRecipient: Text;
+    begin
+        if Recipient = '' then
```
Comment (mazhelez) [inline Modules/System/Email/src/Message/EmailMessageImpl.Codeunit.al:368]: Why not use `SetRecipients`. Like...

```
        Recipient := DelChr(Recipient, '<>'); // trim the whitespaces around

        if Recipient = '' then
            exit;

        Recipient := Recipient.ToLower();

        Recipients := GetRecipients(RecipientType);

        if Recipients.Contains(Recipient) then
            exit;

        Recipients.Add(Recipient);
        SetRecipients(RecipientType, Recipients);
 ```

File: Modules/System Tests/Email/src/EmailMessageUnitTest.Codeunit.al [hunk 26]
```diff
@@ -522,6 +522,279 @@ codeunit 134689 "Email Message Unit Test"
...
+        Index: Integer;
+    begin
+        // Initialize
+        PermissionsMock.Set('Email Edit');
+
+        // Exercise
+        Message.Create(Recipients, 'Test subject', 'Test body', true);
+        Message.SetBodyHTMLFormatted(false);
+        // Verify
+        Assert.IsFalse(Message.IsBodyHTMLFormatted(), 'Body was expected not to be HTML formatted');
```
Comment (mazhelez) [inline Modules/System Tests/Email/src/EmailMessageUnitTest.Codeunit.al:613]: Might be worth checking that the subject and the body are unchanged.

**Review by mazhelez — APPROVED** [review]
Looks good :)

Feel free to address the comments I left. In any case, I'll start the uptake.

**Comment by JesperSchulz** [comment]
Thanks for reporting this. We agree, and we’ll publish a fix asap, either in an update for the current version or in the next major release. Please do not reply to this, as we do not monitor closed issues. If you have follow-up questions or requests, please create a new issue where you reference this one.

**Comment by JesperSchulz** [comment]
Availability update: We will publish a fix for this issue in the next update for release 20.
